File: tools/merge_pmw_message_sql.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Merge two phpMyAdmin-style INSERT dumps for `pmw_message` without dropping rows.

The second file's `id` values are shifted so every new id is > max(id) in the
first file, avoiding duplicate primary keys on import.

Usage:
  python tools/merge_pmw_message_sql.py ^
    "C:\\Users\\niego\\Downloads\\pmw_message (1).sql" ^
    "C:\\Users\\niego\\Downloads\\pmw_message.sql" ^
    -o "C:\\Users\\niego\\Downloads\\pmw_message_merged.sql"

Then import `pmw_message_merged.sql` in MySQL (phpMyAdmin / mysql CLI).

If the table already has rows (#1062 duplicate PRIMARY), pass the current max id:
  python tools/merge_pmw_message_sql.py ... -o out.sql -b 5000
  (-b / --ensure-above-id: shift ALL ids so min(id) > N; use SELECT MAX(id) FROM pmw_message;)
"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

VALUE_LINE = re.compile(r"^\((\d+)\s*,")


def read_text(path: Path) -> str:
    for enc in ("utf-8-sig", "utf-8", "gbk"):
        try:
            return path.read_text(encoding=enc)
        except UnicodeDecodeError:
            continue
    return path.read_text(encoding="utf-8", errors="replace")


def extract_insert_parts(text: str) -> tuple[str, list[str]]:
    """Return (header line ending with VALUES, list of value lines without header)."""
    lines = text.splitlines()
    header_idx = None
    for i, line in enumerate(lines):
        if "INSERT INTO" in line and "pmw_message" in line and "VALUES" in line:
            header_idx = i
            break
    if header_idx is None:
        raise ValueError("Could not find INSERT INTO `pmw_message` ... VALUES in file")

    header = lines[header_idx].strip()
    value_lines: list[str] = []
    for line in lines[header_idx + 1 :]:
        s = line.strip()
        if not s or s.startswith("--"):
            continue
        if VALUE_LINE.match(s):
            value_lines.append(s)
    if not value_lines:
        raise ValueError("No value rows found after INSERT header")
    return header, value_lines


def ids_from_value_lines(value_lines: list[str]) -> list[int]:
    out: list[int] = []
    for s in value_lines:
        m = VALUE_LINE.match(s)
        if m:
            out.append(int(m.group(1)))
    return out


def remap_line_leading_id(line: str, offset: int) -> str:
    m = VALUE_LINE.match(line.strip())
    if not m:
        return line
    old = int(m.group(1))
    new_id = old + offset
    stripped = line.strip()
    return VALUE_LINE.sub(f"({new_id},", stripped, count=1)


def ensure_comma_not_semicolon(s: str) -> str:
    s = s.rstrip()
    if s.endswith(");"):
        return s[:-2] + "),"
    if s.endswith(","):
        return s
    return s + ","


def ensure_semicolon(s: str) -> str:
    s = s.rstrip().rstrip(",")
    if s.endswith(");"):
        return s
    if s.endswith(")"):
        return s + ";"
    return s + ";"


def apply_id_shift_all(lines: list[str], extra_offset: int) -> list[str]:
    if extra_offset == 0:
        return lines
    n = len(lines)
    out: list[str] = []
    for i, row in enumerate(lines):
        mapped = remap_line_leading_id(row, extra_offset)
        out.append(
            ensure_comma_not_semicolon(mapped) if i < n - 1 else ensure_semicolon(mapped)
        )
    return out


def merge_sql(
    primary_path: Path, secondary_path: Path, ensure_above_id: int | None
) -> str:
    h1, v1 = extract_insert_parts(read_text(primary_path))
    _, v2 = extract_insert_parts(read_text(secondary_path))

    ids1 = ids_from_value_lines(v1)
    ids2 = ids_from_value_lines(v2)
    max1 = max(ids1)
    min2 = min(ids2)
    offset = max(0, max1 - min2 + 1)

    body: list[str] = []
    for row in v1:
        body.append(ensure_comma_not_semicolon(row.strip()))

    for i, row in enumerate(v2):
        mapped = remap_line_leading_id(row, offset)
        body.append(
            ensure_comma_not_semicolon(mapped)
            if i < len(v2) - 1
            else ensure_semicolon(mapped.rstrip().rstrip(","))
        )

    all_ids = ids1 + [x + offset for x in ids2]
    min_all, max_all = min(all_ids), max(all_ids)
    report = (
        f"-- merge_pmw_message_sql: primary rows={len(v1)} max_id={max1}, "
        f"secondary rows={len(v2)} min_id={min2} max_id={max(ids2)}, "
        f"id_offset applied to secondary={offset}, merged id range {min_all}..{max_all}"
    )

    extra = 0
    if ensure_above_id is not None and ensure_above_id >= 0:
        merged_ids = ids_from_value_lines(body)
        min_m = min(merged_ids)
        extra = max(0, ensure_above_id - min_m + 1)
        if extra > 0:
            body = apply_id_shift_all(body, extra)
            fid = ids_from_value_lines(body)
            report += (
                f", ensure_above_id={ensure_above_id}, extra_shift={extra}, "
                f"final id range {min(fid)}..{max(fid)}"
            )
        else:
            report += f", ensure_above_id={ensure_above_id} (no extra shift)"

    parts = [report, "", h1, ""] + body + [""]
    return "\n".join(parts)


def main() -> int:
    ap = argparse.ArgumentParser(description="Merge two pmw_message INSERT dumps without id collisions.")
    ap.add_argument("primary_sql", type=Path, help="First file: ids are kept as-is")
    ap.add_argument("secondary_sql", type=Path, help="Second file: ids are shifted up if needed")
    ap.add_argument("-o", "--output", type=Path, required=True, help="Output .sql path")
    ap.add_argument(
        "-b",
        "--ensure-above-id",
        type=int,
        default=None,
        metavar="N",
        help="Shift ALL ids so min(id) > N (use SELECT MAX(id) FROM pmw_message)",
    )
    args = ap.parse_args()

    merged = merge_sql(args.primary_sql, args.secondary_sql, args.ensure_above_id)
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(merged, encoding="utf-8")
    print(f"Wrote {args.output} ({len(merged)} bytes)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/merge_pmw_message_sql.php
<?php
/**
 * Merge two phpMyAdmin-style INSERT dumps for `pmw_message` without dropping rows.
 * Shifts `id` in the second file so all new ids are > max(id) in the first file.
 *
 * Usage:
 *   php tools/merge_pmw_message_sql.php ^
 *     "C:\Users\niego\Downloads\pmw_message (1).sql" ^
 *     "C:\Users\niego\Downloads\pmw_message.sql" ^
 *     -o "C:\Users\niego\Downloads\pmw_message_merged.sql"
 *
 * If the table already has rows (duplicate PRIMARY #1062), pass the current
 * maximum id so all new ids sit above it:
 *   php ... -o out.sql -b 5000
 *   (-b is short for --ensure-above-id; use SELECT MAX(id) FROM pmw_message;)
 */

function read_text(string $path): string
{
    $raw = @file_get_contents($path);
    if ($raw === false) {
        fwrite(STDERR, "Cannot read: {$path}\n");
        exit(1);
    }
    if (substr($raw, 0, 3) === "\xEF\xBB\xBF") {
        $raw = substr($raw, 3);
    }
    return $raw;
}

/** @return array{0: string, 1: string[]} */
function extract_insert_parts(string $text): array
{
    $lines = preg_split("/\r\n|\n|\r/", $text);
    $headerIdx = null;
    foreach ($lines as $i => $line) {
        if (strpos($line, 'INSERT INTO') !== false
            && strpos($line, 'pmw_message') !== false
            && strpos($line, 'VALUES') !== false) {
            $headerIdx = $i;
            break;
        }
    }
    if ($headerIdx === null) {
        throw new RuntimeException('Could not find INSERT INTO `pmw_message` ... VALUES');
    }
    $header = trim($lines[$headerIdx]);
    $valueLines = [];
    for ($j = $headerIdx + 1, $n = count($lines); $j < $n; $j++) {
        $s = trim($lines[$j]);
        if ($s === '' || strpos($s, '--') === 0) {
            continue;
        }
        if (preg_match('/^\(\d+\s*,/', $s)) {
            $valueLines[] = $s;
        }
    }
    if ($valueLines === []) {
        throw new RuntimeException('No value rows found after INSERT header');
    }
    return [$header, $valueLines];
}

/** @param string[] $valueLines */
function ids_from_value_lines(array $valueLines): array
{
    $ids = [];
    foreach ($valueLines as $s) {
        if (preg_match('/^\((\d+)\s*,/', $s, $m)) {
            $ids[] = (int) $m[1];
        }
    }
    return $ids;
}

function remap_line_leading_id(string $line, int $offset): string
{
    $s = trim($line);
    if (!preg_match('/^\((\d+)\s*,/', $s, $m)) {
        return $line;
    }
    $newId = (int) $m[1] + $offset;
    return preg_replace('/^\(\d+\s*,/', '(' . $newId . ',', $s, 1);
}

function ensure_comma_not_semicolon(string $s): string
{
    $s = rtrim($s);
    if (substr($s, -2) === ');') {
        return substr($s, 0, -2) . '),';
    }
    if (substr($s, -1) === ',') {
        return $s;
    }
    return $s . ',';
}

function ensure_semicolon(string $s): string
{
    $s = rtrim($s);
    $s = rtrim($s, ',');
    if (substr($s, -2) === ');') {
        return $s;
    }
    if (substr($s, -1) === ')') {
        return $s . ';';
    }
    return $s . ';';
}

/** @param string[] $lines */
function apply_id_shift_all(array $lines, int $extraOffset): array
{
    if ($extraOffset === 0) {
        return $lines;
    }
    $out = [];
    $n = count($lines);
    foreach ($lines as $i => $row) {
        $mapped = remap_line_leading_id($row, $extraOffset);
        $out[] = $i < $n - 1 ? ensure_comma_not_semicolon($mapped) : ensure_semicolon($mapped);
    }
    return $out;
}

/**
 * @param string[] $v1
 * @param string[] $v2
 * @return array{0: string, 1: string[], 2: string} header, body lines (last ends with ;), comment line
 */
function merge_body_lines(string $h1, array $v1, array $v2): array
{
    $ids1 = ids_from_value_lines($v1);
    $ids2 = ids_from_value_lines($v2);
    $max1 = max($ids1);
    $min2 = min($ids2);
    $offset = max(0, $max1 - $min2 + 1);

    $body = [];
    $n1 = count($v1);
    foreach ($v1 as $i => $row) {
        $line = trim($row);
        $body[] = ensure_comma_not_semicolon($line);
    }

    $n2 = count($v2);
    foreach ($v2 as $i => $row) {
        $mapped = remap_line_leading_id($row, $offset);
        $body[] = $i < $n2 - 1 ? ensure_comma_not_semicolon($mapped) : ensure_semicolon($mapped);
    }

    $allIds = ids_from_value_lines($v1);
    foreach (ids_from_value_lines($v2) as $oid) {
        $allIds[] = $oid + $offset;
    }
    $minAll = min($allIds);
    $maxAll = max($allIds);

    $comment = '-- merge_pmw_message_sql.php: primary rows=' . count($v1) . ' max_id=' . $max1
        . ', secondary rows=' . count($v2) . ' min_id=' . $min2 . ' max_id=' . max($ids2)
        . ', id_offset applied to secondary=' . $offset
        . ', merged id range ' . $minAll . '..' . $maxAll;

    return [$h1, $body, $comment];
}

/** @param string[] $bodyLines */
function merge_sql_final(string $h1, array $bodyLines, string $comment, ?int $ensureAboveId): string
{
    $extra = 0;
    if ($ensureAboveId !== null && $ensureAboveId >= 0) {
        $ids = ids_from_value_lines($bodyLines);
        $minMerged = min($ids);
        $extra = max(0, $ensureAboveId - $minMerged + 1);
        if ($extra > 0) {
            $bodyLines = apply_id_shift_all($bodyLines, $extra);
            $ids2 = ids_from_value_lines($bodyLines);
            $comment .= ', ensure_above_id=' . $ensureAboveId . ', extra_shift=' . $extra
                . ', final id range ' . min($ids2) . '..' . max($ids2);
        } else {
            $comment .= ', ensure_above_id=' . $ensureAboveId . ' (no extra shift; merged min already higher)';
        }
    }

    $out = [$comment, '', $h1, ''];
    foreach ($bodyLines as $line) {
        $out[] = $line;
    }
    $out[] = '';

    return implode("\n", $out);
}

// --- CLI ---
$argv = $_SERVER['argv'] ?? [];
if (count($argv) < 5) {
    fwrite(STDERR, "Usage: php merge_pmw_message_sql.php <primary.sql> <secondary.sql> -o <out.sql> [-b N]\n");
    fwrite(STDERR, "  -b N / --ensure-above-id N   shift ALL ids so min(id) > N (use SELECT MAX(id) FROM pmw_message)\n");
    exit(1);
}

$primary = $argv[1];
$secondary = $argv[2];
$outPath = null;
$ensureAboveId = null;
for ($i = 3; $i < count($argv); $i++) {
    if ($argv[$i] === '-o' && isset($argv[$i + 1])) {
        $outPath = $argv[$i + 1];
        continue;
    }
    if (($argv[$i] === '-b' || $argv[$i] === '--ensure-above-id') && isset($argv[$i + 1])) {
        $ensureAboveId = (int) $argv[$i + 1];
        if ($ensureAboveId < 0) {
            fwrite(STDERR, "ensure-above-id must be >= 0\n");
            exit(1);
        }
        continue;
    }
}
if ($outPath === null) {
    fwrite(STDERR, "Missing -o output path\n");
    exit(1);
}

try {
    [$h1, $v1] = extract_insert_parts(read_text($primary));
    [, $v2] = extract_insert_parts(read_text($secondary));
    [$h1, $body, $comment] = merge_body_lines($h1, $v1, $v2);
    $merged = merge_sql_final($h1, $body, $comment, $ensureAboveId);
} catch (Throwable $e) {
    fwrite(STDERR, $e->getMessage() . "\n");
    exit(1);
}

$dir = dirname($outPath);
if ($dir !== '' && $dir !== '.' && !is_dir($dir)) {
    @mkdir($dir, 0777, true);
}
if (file_put_contents($outPath, $merged) === false) {
    fwrite(STDERR, "Cannot write: {$outPath}\n");
    exit(1);
}

fwrite(STDOUT, 'Wrote ' . $outPath . ' (' . strlen($merged) . " bytes)\n");
